lib/IDE/ExprCompletion.cpp

```diff
@@ -77,7 +77,12 @@ void ExprTypeCheckCompletionCallback::sawSolutionImpl(
     const constraints::Solution &S) {
   auto &CS = S.getConstraintSystem();
 
-  Type ExpectedTy = getTypeForCompletion(S, CompletionExpr);
+  Type ExpectedTy;
+  if (auto ContextualType = CS.getContextualType(CompletionExpr, /*forConstraint=*/false)) {
+    ExpectedTy = ContextualType;
+  } else {
+    ExpectedTy = getTypeForCompletion(S, CompletionExpr);
+  }
```

Comment:
You'll want to query the solution for the contextual type instead of the constraint system. I'm curious why this is needed though, shouldn't `getTypeForCompletion` get the right type here?

Reply:
Good comment with querying the solution instead of the constraint system.

This is needed in the following case:

```swift
protocol MyProto {}
class MyClass {}

enum Factory {
  static func makeMyClass() -> MyClass { return MyClass() }
}

func testGenericReturn() -> some MyClass & MyProto {
  return Factory.#^COMPLETE^#
}
```

The type for completion ends up being `MyClass`, which makes us report `makeMyClass` with a `Convertible` type relation. In cases where the contextual type is explicitly declared, we just want to get that type without any fixes applied. Does this make sense to you?

<details><summary>-debug-constraints output</summary>

```
+ /Users/alex/sbin/swift-ide-test -code-completion -source-filename /private/tmp/a.swift -code-completion-token COMPLETE -debug-constraints
--- Code Completion ---

---Initial constraints for the given expression---
(code_completion_expr implicit type='$T0' location=/private/tmp/a.swift:9:18 range=[/private/tmp/a.swift:9:10 - line:9:18]
  (type_expr type='Factory.Type' location=/private/tmp/a.swift:9:10 range=[/private/tmp/a.swift:9:10 - line:9:10] typerepr='Factory'))

Score: <default 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>
Contextual Type: some MyClass & MyProto at CodeCompletion@/private/tmp/a.swift:9:18
Type Variables:
  $T0 [allows bindings to: lvalue, noescape, hole] [attributes: hole, potentially_incomplete, delayed] [involves_type_vars: $T1] [#defaultable_bindings: 1] [with possible bindings: <empty>] @ locator@0x1130b7618 [CodeCompletion@/private/tmp/a.swift:9:18]
  $T1 [allows bindings to: hole] [attributes: potentially_incomplete] [involves_type_vars: $T0] [with possible bindings: (subtypes of) MyClass] @ locator@0x1130b76b0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → generic parameter 'τ_0_0']
Inactive Constraints:
  $T1 conforms to _Copyable @ locator@0x1130b76b0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → generic parameter 'τ_0_0']
  $T1 subtype MyClass @ locator@0x1130b77e0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #0 (superclass)]
  $T1 conforms to AnyObject @ locator@0x1130b77e0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #0 (superclass)]
  $T1 conforms to MyProto @ locator@0x1130b7908 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #1 (conformance)]
  $T0 equal $T1 @ locator@0x1130b7660 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type]
Opened types:
  locator@0x1130b7688 [ → opened opaque archetype] opens 'τ_0_0' (τ_0_0) -> $T1

  (Potential Binding(s): 
    ($T1 [allows bindings to: hole] [attributes: potentially_incomplete] [involves_type_vars: $T0] [with possible bindings: (subtypes of) MyClass])
  )
  (attempting type variable binding $T1 := MyClass
    (considering: $T1 conforms to _Copyable @ locator@0x1130b76b0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → generic parameter 'τ_0_0']
      (simplification result:
        (removed constraint: $T1 conforms to _Copyable @ locator@0x1130b76b0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → generic parameter 'τ_0_0'])
      )
      (outcome: simplified)
    )
    (considering: $T1 subtype MyClass @ locator@0x1130b77e0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #0 (superclass)]
      (simplification result:
        (removed constraint: $T1 subtype MyClass @ locator@0x1130b77e0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #0 (superclass)])
      )
      (outcome: simplified)
    )
    (considering: $T1 conforms to AnyObject @ locator@0x1130b77e0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #0 (superclass)]
      (simplification result:
        (removed constraint: $T1 conforms to AnyObject @ locator@0x1130b77e0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #0 (superclass)])
      )
      (outcome: simplified)
    )
    (considering: $T1 conforms to MyProto @ locator@0x1130b7908 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #1 (conformance)]
      (simplification result:
    (attempting fix [fix: add missing protocol conformance] @ locator@0x1130b7908 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #1 (conformance)])
    (increasing 'applied fix' score by 1 @ locator@0x1130b7908 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #1 (conformance)])
        (removed constraint: $T1 conforms to MyProto @ locator@0x1130b7908 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #1 (conformance)])
      )
      (outcome: simplified)
    )
    (considering: $T0 equal $T1 @ locator@0x1130b7660 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type]
      (simplification result:
      )
      (outcome: unsolved)
    )
    (Changes:
      (Newly Bound: 
        > $T1 := MyClass
      )
      (Removed Constraints: 
        > $T1 conforms to _Copyable @ locator@0x1130b76b0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → generic parameter 'τ_0_0']
        > $T1 subtype MyClass @ locator@0x1130b77e0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #0 (superclass)]
        > $T1 conforms to AnyObject @ locator@0x1130b77e0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #0 (superclass)]
        > $T1 conforms to MyProto @ locator@0x1130b7908 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #1 (conformance)]
      )
    )
    (Potential Binding(s): 
      ($T0 [allows bindings to: lvalue, noescape, hole] [with possible bindings: (subtypes of) MyClass])
    )
    (attempting type variable binding $T0 := MyClass
      (considering: $T0 equal $T1 @ locator@0x1130b7660 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type]
        (simplification result:
          (removed constraint: $T0 equal $T1 @ locator@0x1130b7660 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type])
        )
        (outcome: simplified)
      )
      (Changes:
        (Newly Bound: 
          > $T0 := MyClass
        )
        (Removed Constraint: 
          > $T0 equal $T1 @ locator@0x1130b7660 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type]
        )
      )
      (found solution: [component: applied fix(s), value: 1])
    )
  )
--- Discovered 1 solutions ---
--- Solution ---
Fixed score: [component: applied fix(s), value: 1]
Type variables:
  $T0 as MyClass @ locator@0x1130b7618 [CodeCompletion@/private/tmp/a.swift:9:18]
  $T1 as MyClass @ locator@0x1130b76b0 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → generic parameter 'τ_0_0']
Opened types:
  locator@0x1130b7688 [ → opened opaque archetype] opens 'τ_0_0' (τ_0_0) -> MyClass [from $T1]
Fixes:
  [fix: add missing protocol conformance] @ locator@0x1130b7908 [CodeCompletion@/private/tmp/a.swift:9:18 → contextual type → opened opaque archetype → opened generic → type parameter requirement #1 (conformance)]
(failed constraint () -> MyClass conv MyClass @ locator@0x130020200 [])

---Solver statistics---
Total number of scopes explored: 0
Maximum depth reached while exploring solutions: 0
found code completion token COMPLETE at offset 186
Begin completions, 3 items
Keyword[self]/CurrNominal:          self[#Factory.Type#]; name=self
Keyword/CurrNominal:                Type[#Factory.Type#]; name=Type
Decl[StaticMethod]/CurrNominal/TypeRelation[Convertible]: makeMyClass()[#MyClass#]; name=makeMyClass()
End completions
LookedupTypeNames: ['swift_ide_test.Factory', 'Swift.Sendable']

```

</details>

Reply:
I see, should this be the default behavior of `getTypeForCompletion` so it would also handle e.g the following?

```swift
protocol MyProto {}
class MyClass {}

extension MyClass {
  static func makeMyClass() -> MyClass { return MyClass() }
}

func testGenericReturn() -> some MyClass & MyProto {
  return .#^CC^#
}
```

Or do cases like this need to be handled differently?

Reply:
It would make sense to me to fold this logic into `getTypeForCompletion`.

Reply:
Good idea, makes sense. It required a minor other test update where I didn’t feel strongly about the results.